File: Python /TerminalLib/TerminalLib/Terminal.py
import sys
import time
import os
import requests
import shutil
import re

#
# CLEAR TERMINAL
#

def Clear_all():
    print("\033[H\033[2J", flush=True)

#
# WAIT ENTER AND CLEAR
#

def wait_enter_clear(prompt="\n\nPress Enter..."):
    input(prompt)
    print("\033[F\033[K", end="")

#
# COLORS
#

class Colors():
    RESET = "\u001B[0m";
    BLACK = "\u001B[30m";
    RED = "\u001B[31m";
    GREEN = "\u001B[32m";
    YELLOW = "\u001B[33m";
    BLUE = "\u001B[34m";
    PURPLE = "\u001B[35m";
    CYAN = "\u001B[36m";
    WHITE = "\u001B[37m";

#
# READ RESPONSE FROM USER
# 

def read_int(Min_value: int,max_value: int) -> int:
    while True:
        user_input = input()

        if not user_input.strip().isdigit():
            sys.stdout.write("\033[F")
            sys.stdout.write("\033[K")
            continue

        choice = int(user_input)

        if choice <= Min_value - 1:
            sys.stdout.write("\033[F")
            sys.stdout.write("\033[K")
            continue

        if choice >= max_value + 1:
            sys.stdout.write("\033[F")
            sys.stdout.write("\033[K")
            continue

        return choice
    

#    
# TYPEWRITE
#
     
def typewrite(text, speed=0.03):
    for char in text:
        sys.stdout.write(char)
        sys.stdout.flush()
        time.sleep(speed)
    print()  


# 
# BAR
#

def print_progress_bar(current, total, bar_length=20):

    progress = current / total
    filled_length = int(bar_length * progress)
    bar = '∎' * filled_length + '.' * (bar_length - filled_length)
    por = current / total * 100;
    sys.stdout.write(f'\r[{bar}] {por:.2f}% {current}/{total} :) ')
    sys.stdout.flush()
    if current == total:
        print()

#
# PRINT CENTRALIZED TEXT
#

def get_terminal_width():
    return os.get_terminal_size().columns

def strip_ansi_codes(text):
    ansi_escape = re.compile(r'\x1B\[[0-9;]*[A-Za-z]')
    return ansi_escape.sub('', text)

def print_centralizedText(text):
    terminal_width = get_terminal_width()
    lines = text.split('\n')
    
    for line in lines:

        clean_line = strip_ansi_codes(line)
        line_length = len(clean_line)

        padding = max(0, (terminal_width - line_length) // 2)
 
        print(' ' * padding + line)

        

File: Ruby/Terminal.rb
require "io/console"

# -------------------------------------------------
# UTILS
# -------------------------------------------------

def tty?
  STDOUT.tty?
end

def reset_style
  print "\e[0m"
  STDOUT.flush
end

# -------------------------------------------------
# CLEAR TERMINAL
# -------------------------------------------------

def clear_all
  return unless tty?
  print "\e[H\e[2J"
end

# -------------------------------------------------
# WAIT ENTER AND CLEAR
# -------------------------------------------------

def wait_enter_clear(prompt = "\n\nPress Enter...")
  print prompt
  STDIN.gets
  if tty?
    print "\e[F\e[K"
    STDOUT.flush
  end
end

# -------------------------------------------------
# COLORS
# -------------------------------------------------

module Colors
  RESET  = "\e[0m"
  BLACK  = "\e[30m"
  RED    = "\e[31m"
  GREEN  = "\e[32m"
  YELLOW = "\e[33m"
  BLUE   = "\e[34m"
  PURPLE = "\e[35m"
  CYAN   = "\e[36m"
  WHITE  = "\e[37m"
end

# -------------------------------------------------
# READ INTEGER INPUT
# -------------------------------------------------

def read_int(min, max)
  raise "read_int requires an interactive terminal" unless STDIN.tty?

  loop do
    input = STDIN.gets&.strip
    unless input&.match?(/^\d+$/)
      clear_last_line
      next
    end

    value = input.to_i
    if value < min || value > max
      clear_last_line
      next
    end

    return value
  end
end

def clear_last_line
  print "\e[F\e[K"
  STDOUT.flush
end

# -------------------------------------------------
# TYPEWRITE
# -------------------------------------------------

def typewrite(text, speed = 0.03)
  text.each_char do |c|
    print c
    STDOUT.flush
    sleep speed
  end
  puts
end

# -------------------------------------------------
# PROGRESS BAR
# -------------------------------------------------

def print_progress_bar(current, total, bar_length = 20)
  return unless tty?
  return if total <= 0

  progress = [current.to_f / total, 1.0].min
  filled = (bar_length * progress).to_i
  bar = "∎" * filled + "." * (bar_length - filled)
  percent = progress * 100

  print "\r[#{bar}] #{format('%.2f', percent)}% #{current}/#{total}"
  STDOUT.flush

  puts if current >= total
end

# -------------------------------------------------
# CENTERED TEXT
# -------------------------------------------------

def print_centered_text(text)
  unless tty?
    puts text
    return
  end

  width = IO.console.winsize[1]
  text.split("\n").each do |line|
    clean = line.gsub(/\e\[[0-9;]*[A-Za-z]/, "")
    padding = [(width - clean.length) / 2, 0].max
    puts " " * padding + line
  end
end

# -------------------------------------------------
# RAINBOW LOADING
# -------------------------------------------------

def rainbow_loading(text, max_dots, stop, speed = 0.3)
  return unless tty?

  colors = [
    Colors::RED,
    Colors::YELLOW,
    Colors::GREEN,
    Colors::CYAN,
    Colors::BLUE,
    Colors::PURPLE
  ]

  dots = 1
  color_index = 0

  begin
    until stop.call
      print "\r#{colors[color_index]}#{text} #{'.' * dots}#{Colors::RESET}"
      STDOUT.flush
      sleep speed

      dots += 1
      dots = 1 if dots > max_dots
      color_index = (color_index + 1) % colors.size
    end
  ensure
    reset_style
    puts
  end
end

# -------------------------------------------------
# SPINNER
# -------------------------------------------------

def spin_bar(stop, speed = 0.1)
  return unless tty?

  bars = ["|", "/", "-", "\\"]
  index = 0

  begin
    until stop.call
      print "\r#{bars[index]}"
      STDOUT.flush
      sleep speed
      index = (index + 1) % bars.size
    end
  ensure
    reset_style
    puts
  end
end
